packages/contracts/contracts/rbac.clar

```diff
@@ -0,0 +1,169 @@
+    (asserts! (> u128 role) (err "Role is out of range 0-127"))
+    (asserts! (has-role user role) (err "Role already revoked"))
+    (let
+      (
+        (currentRoles (get-roles user))
+        (revokedRole (pow u2 role))
+      )
+      (map-set UserRoles
+        {user: user}
+        {roles: (- currentRoles revokedRole)}
+      )
+      (ok true)
+    )
+  )
+)
+
+;; Returns all roles that possess a specific permission
+;; Returns u0 if permission has not been granted to any role (unknown permission)
+(define-read-only (get-permitted-roles (permission (string-ascii 50)))
+  (default-to u0 (get roles (map-get? PermissionRoles {permission: permission})))
+)
+
+;; Checks if a role has a specific permission
+(define-read-only (has-permission (role uint) (permission (string-ascii 50)))
+  (begin 
+    (asserts! (> u128 role) false)
+    (let
+      (
+        (currentRoles (get-permitted-roles permission))
+        (searchRole (pow u2 role))
+      )
+      (> currentRoles (xor currentRoles searchRole))
+    )
+  )
+)
+
+;; Grant a permission to a role.
+;; Throws an error if :
+;; - caller doesn't have role with granted "grant-permission" permission, 
+;; - provided role is not in range 0-127
+;; - permission has already been granted to the specified role
+(define-public (grant-permission (permission (string-ascii 50)) (role uint))
+  (begin
+    (asserts! (can-execute tx-sender "grant-permission") (err "Unauthorized"))
+    (asserts! (> u128 role) (err "Role is out of range 0-127"))
+    (asserts! (not (has-permission role permission)) (err "Permission already granted"))
+    (let
+      (
+        (currentRoles (get-permitted-roles permission))
+        (newRole (pow u2 role))
+      )
+      (map-set PermissionRoles
+        {permission: permission}
+        {roles: (+ currentRoles newRole)}
+      )
+      (ok true)
+    )
+  )
+)
+
+;; Revoke a permission from a role.
+;; Throws an error if :
+;; - caller doesn't have role with granted "revoke-permission" permission, 
+;; - provided role is not in range 0-127
+;; - permission does not exist for role
+(define-public (revoke-permission (permission (string-ascii 50)) (role uint))
+  (begin
+    (asserts! (can-execute tx-sender "revoke-permission") (err "Unauthorized"))
+    (asserts! (> u128 role) (err "Role is out of range 0-127"))
+    (asserts! (has-permission role permission) (err "The permission does not exist for the specified role."))
+    (let
+      (
+        (currentRoles (get-permitted-roles permission))
+        (revokedRole (pow u2 role))
+      )
+      (map-set PermissionRoles
+        {permission: permission}
+        {roles: (- currentRoles revokedRole)}
+      )
+      (ok true)
+    )
+  )
+)
+
+;; Checks if user has at least one role granted with a specified permission.
+(define-read-only (can-execute (user principal) (permission (string-ascii 50)))
+  (begin
+    (asserts! (not (is-eq CONTRACT_OWNER user)) true)
```

Comment:
We could remove the super powers from the contract_owner by removing the if clause and explicitly granting permission in the contract code like:

```
;; set define role contract owner
(define-constant ROLE_CONTRACT_OWNER (pow u2 u127))
(map-set UserRoles {user: tx-sender} {roles: ROLE_CONTRACT_OWNER})
;; grant permissions for contract owner
(map-set PermissionRoles {permission: "grant-permission"} {roles: ROLE_CONTRACT_OWNER})
(map-set PermissionRoles {permission: "revoke-permission"} {roles: ROLE_CONTRACT_OWNER})
(map-set PermissionRoles {permission: "grant-role"} {roles: ROLE_CONTRACT_OWNER})
(map-set PermissionRoles {permission: "revoke-role"} {roles: ROLE_CONTRACT_OWNER})
```

Reply:
I've found that `asserts!` is quite good choice if we want to implement return early pattern. 
In this case replacing `asserts!` with if statement will not introduce additional nesting so I'm ok with this change.

@pxydn @JMKJR @falkonprods @cryptopanter  what do you think?

Reply:
You want to tell me that we can write `map-set` outside a function and it will be called? Will it be executed during contract initialization?

If we can do that - then I'll be more than happy to strip all super powers from the contract owner.

Reply:
> @pxydn ... what do you think?

i agree with @friedger. this is good with me

Reply:
> If we can do that - then I'll be more than happy to strip all super powers from the contract owner.

Yes, all top level statements are executed during deployment. Prominent example: https://explorer.stacks.co/txid/SP000000000000000000002Q6VF78.genesis?chain=mainnet

Reply:
> I've found that asserts! is quite good choice if we want to implement return early pattern.

I agree in general.